### PR TITLE
DRILL-4560: When new bits register, invoke listeners in ZKClusterCoordinator

### PR DESCRIPTION
**Pull request by paul-rogers**
DRILL-4560: When new bits register, invoke listeners in ZKClusterCoordinator, (JIRA title: ZKClusterCoordinator does not call DrillbitStatusListener.drillbitRegistered for new bits.) While this functionality is not needed in Drill itself, it is needed for tools that monitor Drill (such as Drill on YARN.) Adding this code to Drill avoids the need to clone the file for other tools.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java [hunk 1]
```diff
@@ -158,10 +158,14 @@ public void cacheChanged() {
     }
   }
 
+  @Override
   public void close() throws Exception {
-    // discovery attempts to close its caches(ie serviceCache) already. however, being good citizens we make sure to
-    // explicitly close serviceCache. Not only that we make sure to close serviceCache before discovery to prevent
-    // double releasing and disallowing jvm to spit bothering warnings. simply put, we are great!
+    // discovery attempts to close its caches(ie serviceCache) already. however,
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: What's changed in the comment? Formatting?

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java [hunk 2]
```diff
@@ -231,31 +235,42 @@ public DrillbitEndpoint apply(ServiceInstance<DrillbitEndpoint> input) {
...
+          }
+          logger.debug(builder.toString());
         }
-        logger.debug(builder.toString());
       }
 
-      // Notify the drillbit listener for newly unregistered bits. For now, we only care when drillbits are down / unregistered.
-      if (! (unregisteredBits.isEmpty()) ) {
+      // Notify the drillbit listener for newly unregistered bits.
+      if (!(unregisteredBits.isEmpty())) {
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Remove extra set of parens (here and below).

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java [hunk 3]
```diff
@@ -231,31 +235,42 @@ public DrillbitEndpoint apply(ServiceInstance<DrillbitEndpoint> input) {
...
+            builder.append(bit.getUserPort());
+            builder.append(':');
+            builder.append(bit.getControlPort());
+            builder.append(':');
+            builder.append(bit.getDataPort());
+            builder.append('\n');
+          }
+          logger.debug(builder.toString());
         }
-        logger.debug(builder.toString());
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Undo this change; client side logs could show an empty drill cluster.

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java (hunk 2 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Correct the comment to "Notify **all listeners** for newly unregistered bits." (similarly correction below).

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java [hunk 4]
```diff
@@ -231,31 +235,42 @@ public DrillbitEndpoint apply(ServiceInstance<DrillbitEndpoint> input) {
...
 
-      // Notify the drillbit listener for newly unregistered bits. For now, we only care when drillbits are down / unregistered.
-      if (! (unregisteredBits.isEmpty()) ) {
+      // Notify the drillbit listener for newly unregistered bits.
+      if (!(unregisteredBits.isEmpty())) {
         drillbitUnregistered(unregisteredBits);
       }
+      // Notify the drillbit listener for newly registered bits.
+      if (!(registeredBits.isEmpty())) {
+        drillbitRegistered(registeredBits);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Are all callbacks empty blocks?

**Comment by sudheeshkatkam** [comment]
Please fix the commit message. Maybe "DRILL-4560: When new bits register, invoke listeners  in ZKClusterCoordinator"?

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java (hunk 1 above)
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Formatting. Not essential, I'll revert the change.

**Comment by paul-rogers** [comment]
Changed PR title.

**Comment by sudheeshkatkam** [comment]
Not the PR title, the commit [message](https://help.github.com/articles/changing-a-commit-message/) :)

**Comment by paul-rogers** [comment]
Sorry, the commit message quotes the JIRA entry title. I'll change the message, but quote the JIRA title in the body.

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java [hunk 5]
```diff
@@ -248,15 +253,18 @@ public DrillbitEndpoint apply(ServiceInstance<DrillbitEndpoint> input) {
           builder.append(':');
           builder.append(bit.getDataPort());
           builder.append('\n');
+          logger.debug(builder.toString());
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Actual logging should be outside the for-loop?

**Comment by sudheeshkatkam** [comment]
+1, pending minor comment

Can you squash the changes as well?

File: exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java [hunk 6]
```diff
@@ -245,33 +242,29 @@ public DrillbitEndpoint apply(ServiceInstance<DrillbitEndpoint> input) {
...
-            builder.append(bit.getAddress());
-            builder.append(':');
-            builder.append(bit.getUserPort());
-            builder.append(':');
-            builder.append(bit.getControlPort());
-            builder.append(':');
-            builder.append(bit.getDataPort());
-            builder.append('\n');
-          }
+        builder.append(" total bits. New active drillbits:\n");
```
Comment (sohami) [inline exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKClusterCoordinator.java]: Please change this log from "**New** active drillbits.." to "**Current** active drillbits"
